### PR TITLE
Add PyTorch-based componentized architecture.

### PR DESCRIPTION
**Pull request by mallamanis**
Add two high-level classes to assist with development of PyTorch models.

* `BaseComponent` is a wrapper around `nn.Module` that enforces a specific structure. Components may have other components as children.
* `ComponentTrainer` that takes a (root) `BaseComponent` and trains the model.

### CODE REVIEW

File: python/dpu_utils/utils/iterators.py [hunk 1]
```diff
@@ -188,3 +189,27 @@ def __next__(self):
             self.__worker_process_outer.join()
             raise StopIteration
         return next_element
+
+
+def shuffled_iterator(input: Iterator[T], buffer_size: int = 10000, out_slice_sizes: int = 500) -> Iterator[T]:
```
Comment (pashminacameron) [inline python/dpu_utils/utils/iterators.py]: `input` -> `input_iterator` might be good, to avoid conflict with Python's `input()`

File: python/dpu_utils/ptutils/componenttrainer.py [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+            num_minibatches = 0
+            num_samples = 0
+            have_more_data = True
+            start_time = time.time()
+            self.__model.reset_metrics()
+            with tqdm(desc='Validation', disable=not show_progress_bar, leave=False) as progress_bar, torch.no_grad():
+                while have_more_data:
+                    mb_data, have_more_data, num_elements = self.__model.create_minibatch(data_iter,
+                                                                                          max_num_items=self.__minibatch_size)
+                    if num_elements == 0:
```
[the remark this answers is not in the document]
Comment (mallamanis) [reply python/dpu_utils/ptutils/componenttrainer.py]: Thanks! It's at least confusing you're right. I've now tried to make things somewhat simpler. Let me know if these make sense or if you see any way to improve this.

The principle I am using here is that the `data_iter` is gradually consumed. When it's entirely consumed then we stop iterating. During training, we skip the last minibatch that will be smaller than the other ones (because it has fewer samples, the gradient will be more noisy). During validation, we still want to account for partial (non-empty) minibatches to compute the validation score.

[the remark this answers is not in the document]
Comment (pashminacameron) [reply python/dpu_utils/ptutils/componenttrainer.py]: Yes, this is much clear. Thanks.
